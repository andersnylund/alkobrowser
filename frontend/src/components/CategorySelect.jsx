import React from 'react';
import { Query } from 'react-apollo';
import gql from 'graphql-tag';
import { Select } from 'semantic-ui-react';
import { useSelector, useDispatch } from 'react-redux';

import { allCategories } from '../constants';
import { setSelectedCategoryAction } from '../actions/filterActions';

const CATEGORIES = gql`
  query {
    categories(orderBy: tyyppi_ASC) {
      id
      tyyppi
    }
  }
`;

const CategorySelect = () => {
  const selectedCategory = useSelector(state => state.filter.selectedCategory);
  const dispatch = useDispatch();

  return (
    <Query query={CATEGORIES}>
      {({ data, loading, error }) => {
        if (loading) {
          return <Select options={[]} />;
        }
        if (error) {
          return null;
        }
        const options = [
          allCategories,
          ...data.categories.map(c => ({
            key: c.id,
            text: c.tyyppi,
            value: c.id,
          })),
        ];
        return (
          <Select
            options={options}
            value={selectedCategory}
            onChange={(event, { value }) =>
              dispatch(setSelectedCategoryAction(value))
            }
          />
        );
      }}
    </Query>
  );
};

export default CategorySelect;
