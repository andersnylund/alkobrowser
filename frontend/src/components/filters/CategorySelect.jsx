import React from 'react';
import { Query } from 'react-apollo';
import gql from 'graphql-tag';
import { Select } from 'semantic-ui-react';
import { connect } from 'react-redux';
import { func, string } from 'prop-types';

import { allCategories } from '../../constants';
import { setSelectedCategoryAction } from '../../actions/filterActions';

export const CATEGORY_QUERY = gql`
  query {
    categories(orderBy: tyyppi_ASC) {
      id
      tyyppi
    }
  }
`;

export const CategorySelect = ({ selectedCategory, setSelectedCategory }) => {
  return (
    <Query query={CATEGORY_QUERY}>
      {({ data, loading, error }) => {
        if (loading) {
          return <Select options={[]} />;
        }
        if (error) {
          return null;
        }
        const options = [
          allCategories,
          ...data.categories.map(c => ({
            key: c.id,
            text: c.tyyppi,
            value: c.id,
          })),
        ];
        return (
          <Select
            options={options}
            value={selectedCategory}
            onChange={(event, { value }) => setSelectedCategory(value)}
          />
        );
      }}
    </Query>
  );
};

CategorySelect.propTypes = {
  selectedCategory: string.isRequired,
  setSelectedCategory: func.isRequired,
};

const mapStateToProps = state => ({
  selectedCategory: state.filter.selectedCategory,
});

const mapDispatchToProps = {
  setSelectedCategory: setSelectedCategoryAction,
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(CategorySelect);
