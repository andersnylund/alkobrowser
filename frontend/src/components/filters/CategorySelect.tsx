import React, { FC, useEffect } from 'react';
import { Select } from 'semantic-ui-react';
import { connect } from 'react-redux';

import { allCategories } from '../../constants';
import { setSelectedCategoryAction } from '../../actions/filterActions';
import { getCategoriesAction } from '../../actions/categoryActions';
import { AppState } from '../../store';
import { Category } from '../../../../shared/types';

export interface Props {
  selectedCategory: string;
  setSelectedCategory: (selectedCategory: string) => void;
  getCategories: () => void;
  categories: Category[];
}

export const CategorySelect: FC<Props> = ({
  selectedCategory,
  setSelectedCategory,
  getCategories,
  categories,
}) => {
  useEffect(() => {
    getCategories();
  }, [getCategories]);

  const options = [
    allCategories,
    ...categories.map(c => ({
      key: c.id,
      text: c.tyyppi,
      value: c.id,
    })),
  ];
  return (
    <Select
      options={options}
      value={selectedCategory}
      onChange={(event, e) => {
        if (typeof e.value === 'string') {
          setSelectedCategory(e.value);
        }
      }}
    />
  );
};

const mapStateToProps = (state: AppState) => ({
  selectedCategory: state.filter.selectedCategory,
  categories: state.category.categories,
});

const mapDispatchToProps = {
  setSelectedCategory: setSelectedCategoryAction,
  getCategories: getCategoriesAction,
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(CategorySelect);
