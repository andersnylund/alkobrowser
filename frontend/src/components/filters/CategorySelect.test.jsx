import React from 'react';
import { cleanup, render } from '@testing-library/react';
import { MockedProvider } from 'react-apollo/test-utils';
import wait from 'waait';

import { CategorySelect, CATEGORY_QUERY } from './CategorySelect';

describe('<CategorySelect />', () => {
  afterEach(cleanup);

  it('should render without an error', () => {
    const setSelectedCategoryMock = jest.fn();
    render(
      <MockedProvider>
        <CategorySelect
          selectedCategory=""
          setSelectedCategory={setSelectedCategoryMock}
        />
      </MockedProvider>
    );
  });

  it('should show an empty select when loading', () => {
    const setSelectedCategoryMock = jest.fn();

    const categoryMock = {
      request: {
        query: CATEGORY_QUERY,
      },
      result: {
        data: {},
      },
    };

    const { container } = render(
      <MockedProvider mocks={[categoryMock]}>
        <CategorySelect
          selectedCategory=""
          setSelectedCategory={setSelectedCategoryMock}
        />
      </MockedProvider>
    );

    expect(container).toMatchSnapshot();
  });

  it('should show an empty select when loading', async () => {
    const setSelectedCategoryMock = jest.fn();

    const categoryMock = {
      request: {
        query: CATEGORY_QUERY,
      },
      error: new Error('Testing error'),
    };

    const { container } = render(
      <MockedProvider mocks={[categoryMock]}>
        <CategorySelect
          selectedCategory=""
          setSelectedCategory={setSelectedCategoryMock}
        />
      </MockedProvider>
    );

    await wait(0);

    expect(container).toMatchSnapshot();
  });

  it('should list the different categories', async () => {
    const setSelectedCategoryMock = jest.fn();

    const categoryMock = {
      request: {
        query: CATEGORY_QUERY,
      },
      result: {
        data: {
          categories: [
            {
              id: '2',
              tyyppi: 'tyyppi2',
            },
            {
              id: '3',
              tyyppi: 'tyyppi3',
            },
          ],
        },
      },
    };

    const { container } = render(
      <MockedProvider mocks={[categoryMock]} addTypename={false}>
        <CategorySelect
          selectedCategory=""
          setSelectedCategory={setSelectedCategoryMock}
        />
      </MockedProvider>
    );

    await wait(0);

    expect(container).toMatchSnapshot();
  });
});
