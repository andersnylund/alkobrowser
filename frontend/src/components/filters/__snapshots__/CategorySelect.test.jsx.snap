// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CategorySelect /> should list the different categories 1`] = `
<div>
  <div
    aria-expanded="false"
    class="ui selection dropdown"
    role="listbox"
    tabindex="0"
  >
    <div
      aria-live="polite"
      class="text"
      role="alert"
    />
    <i
      aria-hidden="true"
      class="dropdown icon"
    />
    <div
      class="menu transition"
    >
      <div
        aria-checked="false"
        aria-selected="true"
        class="selected item"
        role="option"
        style="pointer-events: all;"
      >
        <span
          class="text"
        >
          Kaikki juomatyypit
        </span>
      </div>
      <div
        aria-checked="false"
        aria-selected="false"
        class="item"
        role="option"
        style="pointer-events: all;"
      >
        <span
          class="text"
        >
          tyyppi2
        </span>
      </div>
      <div
        aria-checked="false"
        aria-selected="false"
        class="item"
        role="option"
        style="pointer-events: all;"
      >
        <span
          class="text"
        >
          tyyppi3
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`<CategorySelect /> should show an empty select when loading 1`] = `
<div>
  <div
    aria-expanded="false"
    class="ui selection dropdown"
    role="listbox"
    tabindex="0"
  >
    <div
      aria-live="polite"
      class="text"
      role="alert"
    />
    <i
      aria-hidden="true"
      class="dropdown icon"
    />
    <div
      class="menu transition"
    />
  </div>
</div>
`;

exports[`<CategorySelect /> should show an empty select when loading 2`] = `<div />`;
